### ISSUE TITLE
Poor performance of FWI results

### ISSUE THREAD

**Issue by Lujiang-ECUST** [ar4/deepwave#93]
Hi Alan,
I'm trying to perform FWI on my model, and despite trying many settings the inversion is not good. Even if I set the initial model to be very similar to the real model, the inversion results get worse with iterations. I'd like to know what's going wrong and how can I improve it!
This is the result after performing FWI

https://github.com/user-attachments/assets/aaee11f3-0964-4460-ba38-b9dd42d23921

**Comment by Lujiang-ECUST** [comment]
I'll attach my part of the code
'''

ny = 400
nx = 400
dx = 0.1e-3
vp_background = torch.ones(ny,nx, device=device) *5724
n_shots = 64

n_sources_per_shot = 1
d_source =6
first_source = 8
source_depth = 0

n_receivers_per_shot1 =64
n_receivers_per_shot2 = 32
d_receiver =6
first_receiver = 8
receiver_depth1 = 0
receiver_depth2 = 99
freq = 2.25e6
nt =1000
dt = 2e-8
peak_time =2.25 / freq+75*dt
# source_locations
source_locations = torch.zeros(n_shots, n_sources_per_shot, 2,device=device)
source_locations[..., 0] = source_depth
source_locations[:, 0, 1] = (torch.arange(n_shots) * d_source +
                             first_source)

# receiver_locations
receiver_locations = torch.zeros(n_shots, n_receivers_per_shot1, 2, device=device)
receiver_locations[:,:, 0] = receiver_depth1

receiver_locations[:, :, 1] = (
    (torch.arange(n_receivers_per_shot1) * d_receiver +
     first_receiver)
    .repeat(n_shots, 1)
)
# source_amplitudes
source_amplitudes = (
    (deepwave.wavelets.ricker(freq, nt, dt, peak_time))
    .repeat(n_shots, n_sources_per_shot, 1).to(device)
)
out_base = deepwave.scalar(
                vp_background, dx, dt,
                source_amplitudes=source_amplitudes,
                source_locations=source_locations,
                receiver_locations=receiver_locations,
                pml_freq=freq,
                pml_width=[0,0,20,20],
                time_pad_frac=0.2,
                accuracy=4)[-1]
out_base[:,:,350:1000]=0
observed_data = deepwave.scalar(
                vp_true, dx, dt,
                source_amplitudes=source_amplitudes,
                source_locations=source_locations,
                receiver_locations=receiver_locations,
                pml_freq=freq,
                pml_width=[0,0,20,20],
                time_pad_frac=0.2,
                accuracy=4)[-1]
observed_data=observed_data-out_base
vp = (torch.tensor(1/gaussian_filter(1/vp_true.detach().cpu().numpy(), 10))
          .to(device))
loss_fn = torch.nn.MSELoss()
def taper(x):
    return deepwave.common.cosine_taper_end(x, 100)
# Run optimisation/inversion
n_epochs =10
n_batch=4
n_shots_per_batch = (n_shots + n_batch - 1) // n_batch
observed_data = taper(observed_data)
class Model(torch.nn.Module):
    def __init__(self, initial, min_vel, max_vel):
        super().__init__()
        self.min_vel = min_vel
        self.max_vel = max_vel
        self.model = torch.nn.Parameter(
            torch.logit((initial - min_vel) /
                        (max_vel - min_vel))
        )

    def forward(self):
        return (torch.sigmoid(self.model) *
                (self.max_vel - self.min_vel) +
                self.min_vel)
class Prop(torch.nn.Module):
    def __init__(self, model, dx, dt, freq):
        super().__init__()
        self.model = model
        self.dx = dx
        self.dt = dt
        self.freq = freq

    def forward(self, source_amplitudes, source_locations,
                receiver_locations):
        out = deepwave.scalar(
            model().to(device), self.dx, self.dt,
            source_amplitudes=source_amplitudes.to(device),
            source_locations=source_locations,
            receiver_locations=receiver_locations,
            max_vel=5800,
            pml_freq=self.freq,
            pml_width=[0,0,20,20],
            time_pad_frac=0.2,
        )
        return out[-1]
model = Model(vp, 1400, 5800).to(device)
prop = Prop(model, dx, dt, freq)
prop = torch.nn.DataParallel(prop).to(device)
for cutoff_freq in [1.5e6,2.25e6,2.75e6]:
    sos = butter(6, cutoff_freq, fs=1/dt, output='sos')
    sos = [torch.tensor(sosi).to(observed_data.dtype).to(device)
           for sosi in sos]

    def filt(x):
        return biquad(biquad(biquad(x, *sos[0]), *sos[1]), *sos[2])
    observed_data = filt(observed_data)
    # optimiser = torch.optim.LBFGS(model.parameters(),
    #                               line_search_fn='strong_wolfe')
    optimiser = torch.optim.SGD(model.parameters(), lr=1e2, momentum=0.9)
    for epoch in range(n_epochs):
        optimiser.zero_grad()
        for batch in range(n_batch):
            batch_start = batch * n_shots_per_batch
            batch_end = min(batch_start + n_shots_per_batch,
                                n_shots)
            if batch_end <= batch_start:
                continue
            s = slice(batch_start, batch_end)
            out=prop(source_amplitudes[s], source_locations[s],
                            receiver_locations[s])
            out=out-out_base[s]
            out = filt(taper(out))
            loss = 1e25 * loss_fn(out, observed_data[s])
            loss.backward()
            torch.nn.utils.clip_grad_norm_(
                model.parameters(),max_norm=20,norm_type=2
            )
            optimiser.step()
'''

**Comment by ar4** [comment]
Hello and thank you for your question.

Your code does a lot of complex things, including parallel execution,
frequency filtering, an optimiser with momentum, gradient clipping, and
(with the Model class) using a sigmoid. I imagine that this is probably
because, as you say, you tried many things, but I suggest keeping the setup
as simple as possible and only adding complexity when you are sure it is
necessary. Your approach of using an initial model that is very close to or
even the same as the true model seems wise, and I think the result should
not get worse when doing that if you remove all of the complexity and just
perform simple forward and backpropagation.

I am also concerned about the creation of your observed data. It appears as
if you try to remove the direct arrival. That seems like an unusual thing
to do when performing FWI, and seems certain to cause problems when you
then do not also remove the direct arrival from the forward modelled data
in the inversion.

**Comment by Lujiang-ECUST** [comment]
> 您好，感谢您的提问。您的代码执行了很多复杂的事情，包括并行执行、频率过滤、带动量的优化器、梯度裁剪以及（使用 Model 类）使用 sigmoid 进行。我想这可能是因为，正如您所说，您尝试了很多方法，但我建议保持设置尽可能简单，并且仅在确定 必要。 您使用非常接近或甚至与真实模型相同的初始模型的方法似乎是明智的，我认为结果应该如果您去除所有复杂性并只执行简单的正向和反向传播。我还担心您观察到的数据的创建。它显示为如果您尝试删除直接到达。这似乎是一件不寻常的事情在执行 FWI 时要做。

I followed your advice and simplified the setup. Parallel execution is necessary for me and I only use the model class and the lbfs optimiser for inversion . I'm not sure about the Model class has a significant impact on the results

https://github.com/user-attachments/assets/b961fdfe-5619-41ef-9345-3f081e8c72bf

**Comment by Lujiang-ECUST** [comment]
I have tried the lbfgs optimiser and when I set the initial model to a homogeneous background and perform the inversion at frequencies [1.5e6,2.25e6,2.75e6], the inversion results are as follows, and I am puzzled as to why the model seems to always be updated with a ‘scattering of speckles’.

https://github.com/user-attachments/assets/aba49a8f-aea2-4c54-9dd1-d541a904f4ac

**Comment by Lujiang-ECUST** [comment]
> 此外，除非我弄错了，否则您似乎正在用 Loop 内的 Filtered 版本覆盖 Cutoff 频率。因此，较高频率的 cutoffs 将无法按预期工作，因为您已经过滤掉了较高的频率。`observed_data`
I have used a high frequency filter to achieve better results in a previous case #85 , I am not sure if the filter worked or not, when I inverted the results with only 2.25e6hz the results were still poor

**Comment by Lujiang-ECUST** [comment]
> Also, unless I am mistaken, it looks like you are overwriting your with the filtered version inside the loop over cutoff frequencies. The higher frequency cutoffs will therefore not work as intended as you have already filtered out the higher frequencies.`observed_data`

Oh, I see, you're right, I should have used observed_data_filt

**Comment by ar4** [comment]
Thank you for your patience in trying to get good results with Deepwave.

If you are still not getting results that you are satisfied with, would it be possible for you to send me a full code that I can run? That might help me to identify the problem.

**Comment by Lujiang-ECUST** [comment]
This is my complete code, the file contains the loaded real wave speed model as well as the experimental observations. I plan to get good results on the simulated data first and then invert it on the experimental data（with direct arrivals removed).

[code_to_Alan.zip](https://github.com/user-attachments/files/20325571/code_to_Alan.zip)

**Comment by ar4** [comment]
Thank you for your code.

Your usage of Deepwave looks correct, but I see some other issues with your code.

One of these is that the loss value that you return to the optimiser is only the loss calculated from the last batch. You should instead accumulate the loss value over all of the batches and return that:

```python
def closure():
            optimiser.zero_grad()
            total_loss = 0
            for batch in range(n_batch):
                ...
                loss.backward()
                total_loss += loss.item()
            print(total_loss)
            return total_loss
```

Another is that I recommend specifying the `max_vel` parameter in all of your calls to Deepwave to ensure that the subtraction of the direct arrival that you do is as accurate as possible. Deepwave uses the maximum velocity for purposes such as calculating the internal time step size to use. If you do not specify it, Deepwave will calculate the maximum velocity from the velocity model, but if that varies between calls then it can cause small differences in the output that result in the direct arrival not cancelling well.

You call `closure` at the beginning of each epoch to calculate the loss. That seems wasteful, especially as it will also run backpropagation and then discard the gradients. There are more efficient options, such as saving the loss from the optimiser's calls to `closure` and only printing it at the end of each epoch.

Do you still get the "scattering of speckles" that you reported, or other poor result, when using this code? I only ran it (after fixing the above issues) using one shot and one epoch, but the result looked quite good to me.

I imagine that getting good results on real data will be difficult, and something that I probably do not have enough experience with to be of assistance.

**Comment by Lujiang-ECUST** [comment]
Thank you for your correction. I changed the code as suggested but the result after inversion is still not good. I'm not sure if I changed it correctly so I attached the modified code .
[FWI_deepwave_modified.txt](https://github.com/user-attachments/files/20360855/FWI_deepwave_modified.txt)

Have you tried the inversion with the initial model as a homogeneous model and did it turn out good?

I tried ten epochs with the initial model being similar as well as the homogeneous model, and this is the inversion result.

https://github.com/user-attachments/assets/f13ba75f-21b3-4dad-b100-fc8ea87336cf

https://github.com/user-attachments/assets/c355b3dc-4466-49cf-874c-0344aa52761e
Thanks again for your reply and valuable suggestions. After getting good results on the simulated data, I will try as much as I can on the experimental data and share the results with you!

**Comment by ar4** [comment]
Yes, it looks like you implemented the changes correctly, although, if you want to, you can reduce the frequency of printing the loss by moving the print statement outside of `closure` (but you will then need to write the loss to a variable defined outside `closure`, such as by declaring that variable to be `global` inside `closure`).

I think the poor results problem is mainly caused by the two free surfaces you have. These will cause a lot of strong multiples, which makes inversion difficult. I do not think this is an issue with Deepwave, but instead just an example of why getting good results with FWI is hard. If you make all of the boundaries absorbing, I expect that you will get better results. Since your background model is correct and your anomaly is quite spatially small, you might also consider trying the inversion without frequency filtering.

**Comment by ar4** [comment]
Hello,
I am sorry to hear that you are still having difficulty with this. Do I correctly understand that you now do not remove the direct arrivals, so they are included in the inversion, and you use absorbing boundaries on all sides? In that case my initial guess is that not many direct arrivals are passing through the anomaly and returning to the receivers, so the data does not have a lot of sensitivity to the low wavenumber components of the anomaly, but if you can provide me with the latest version of your code and an image of the result you are getting, I will have a look.

**Comment by ar4** [comment]
Thank you.

I have looked at your code and experimented with it, and you again seem to be using Deepwave correctly, so I think the problem is simply that FWI is difficult and you are trying to apply it to a situation that is particularly difficult.

I tried several ideas that should have made the problem easier, such as converting it to a transmission problem (moving the receivers to the bottom of the model), but the results were still not good. Although I can think of other things to try, I am not certain whether they will help. My advice is to start with an easy problem, get good results, and then gradually make the problem more similar to the one that you really want to apply it to. I suggest starting with a transmission problem (sources on one side of an anomaly, receivers on the other), in a model where the background velocity is 1500 m/s and there is a small anomaly in the centre that has a slightly higher velocity than the background (say 1600 m/s). I also suggest using parameters closer to typical seismic surveys (frequency about 25 Hz, dt = 0.004, dx = 5, etc.) to avoid the risk that unusual parameters cause strange results.

I am sorry that I have not been able to be of more help, but I hope this suggestion sets you on a productive path to finding the solution. Please feel free to keep me updated with your progress and let me know if you have further questions.
